### PR TITLE
feat: lazy load cms images /learn/ pages

### PR DESCRIPTION
**Pull request by fairlighteth**
# Summary

- Adds lazy loading for images coming from the CMS for the /learn/ and all /learn/* pages
- Adds lazy loading for <SVG by using a helper component <LazySVG
- Lighthouse optimizations (color contrast, image dimensions)
- Modified the `sitemap.xml` generating process to set the `lastMod` property for /learn/ articles to use the last modified date from the CMS. For the the /tokens/* pages it assumes daily updates. The other static pages for now will update also daily (on each build).
- Fixed /tokens and /tokens/* meta title/descriptions
- Fixed /learn/* (articles) meta descriptions: They use the `description` field from CMS, with a fallback in place.
- Adds Google Tag Manager by loading along the initAnalytics. It will **not** load by default, unless initialized as `initCowAnalyticsGoogle(true)`. Google Tag Manager may only load on cow.fi and **not** on swap.cow.fi 
- Consolidated the CategoryLinks component

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @fairlighteth , when I open Learn page. I don't see that Topics images are loaded 
image

They appear only when I press on any topic and then go back. Could you please check why this happens?

File: apps/cow-fi/hooks/useLazyLoadImages.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  rootMargin: '10px',
+  placeholderColor: '#f0f0f0',
+  fadeInDuration: '0.3s',
+  minHeight: '100px',
+}
+
+export function useLazyLoadImages() {
+  const replaceImageUrls = React.useCallback((html: string): string => {
+    return html.replace(/<img([^>]*)src="([^"]*)"([^>]*)>/g, (match, before, src, after) => {
+      return `<img${before}data-src="${src}" src="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='100%' height='100%' viewBox='0 0 1 1' preserveAspectRatio='none'%3E%3Crect width='1' height='1' fill='${LAZY_LOADING_CONFIG.placeholderColor.replace('#', '%23')}' /%3E%3C/svg%3E"${after}>`
```
Comment (alfetopito) [inline apps/cow-fi/hooks/useLazyLoadImages.tsx:13]: This part is almost all duplicated with line `50`.
Can you extract this into a fn outside of the hook?

**Review by alfetopito — APPROVED** [review]

File: apps/cow-fi/hooks/useLazyLoadImages.tsx (hunk 1 above)
Comment (fairlighteth) [reply apps/cow-fi/hooks/useLazyLoadImages.tsx]: Addressed this now!

**Comment by fairlighteth** [comment]
> Hey @fairlighteth , when I open Learn page. I don't see that Topics images are loaded image
> 
> They appear only when I press on any topic and then go back. Could you please check why this happens?

Should be fixed 👍🏼

**Comment by fairlighteth** [comment]
@elena-zh @alfetopito did updates to this PR so requesting a new review

**Review by alfetopito — APPROVED** [review]

File: apps/cow-fi/pages/learn/articles/[[...page]].tsx [hunk 2]
```diff
@@ -75,8 +76,45 @@ export type ArticlesResponse = {
...
+          onClick={() => clickOnKnowledgeBase(`click-categories-${category.name}`)}
+        >
+          {category.name}
+        </a>
+      </li>
+    ))}
+  </CategoryLinks>
+)
+
+const renderArticles = (articles: any[]) => (
```
Comment (shoom3301) [inline apps/cow-fi/pages/learn/articles/[[...page]].tsx:99]: It supposed to be a component

File: apps/cow-fi/pages/learn/articles/[[...page]].tsx [hunk 3]
```diff
@@ -75,8 +76,45 @@ export type ArticlesResponse = {
   }
 }
 
+const renderCategoryLinks = (allCategories: { name: string; slug: string }[]) => (
```
Comment (shoom3301) [inline apps/cow-fi/pages/learn/articles/[[...page]].tsx:79]: It supposed to be a component

File: apps/cow-fi/components/LazySVG.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  width,
+  height,
+  ...props
+}) => {
+  const [isInView, setIsInView] = useState(false)
+  const wrapperRef = useRef<HTMLSpanElement>(null)
+
+  useEffect(() => {
+    if (!wrapperRef.current) {
+      console.log('wrapperRef is null')
```
Comment (shoom3301) [inline apps/cow-fi/components/LazySVG.tsx:23]: Should we log it?

File: apps/cow-fi/hooks/useLazyLoadImages.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      const handleIntersect = (entry: IntersectionObserverEntry) => {
+        const img = entry.target as HTMLImageElement
+        if (entry.isIntersecting && img.dataset.src) {
+          img.src = img.dataset.src
+          img.style.opacity = '1'
+          observer?.unobserve(img)
+        }
+      }
+
+      initObserver(handleIntersect)
```
Comment (shoom3301) [inline apps/cow-fi/hooks/useLazyLoadImages.tsx:65]: I think `initObserver` function doesn't work right.
So, it creates `IntersectionObserver` only once regardless of a context.
It means that if you call the `useLazyLoadImages()` hook once, in the next time it won't work and `handleIntersect` won't be called

File: apps/cow-fi/hooks/useLazyLoadImages.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+  observer = new IntersectionObserver(
+    (entries) => {
+      entries.forEach(onIntersect)
+    },
+    { rootMargin: LAZY_LOADING_CONFIG.rootMargin },
+  )
+}
+
+export function useLazyLoadImages() {
```
Comment (shoom3301) [inline apps/cow-fi/hooks/useLazyLoadImages.tsx:31]: Can we just use [native lazy-loading](https://developer.mozilla.org/en-US/docs/Web/Performance/Lazy_loading) instead?

Comment (shoom3301) [reply apps/cow-fi/hooks/useLazyLoadImages.tsx:31]: Or https://github.com/Aljullu/react-lazy-load-image-component

Comment (fairlighteth) [reply apps/cow-fi/hooks/useLazyLoadImages.tsx:31]: The issue was that content was being loaded from markdown/cms and  we need to deal what that content

File: apps/cow-fi/hooks/useLazyLoadImages.tsx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  const placeholderSrc = useRef(getPlaceholderSrc(LAZY_LOADING_CONFIG.placeholderColor)).current
```
Comment (shoom3301) [inline apps/cow-fi/hooks/useLazyLoadImages.tsx:32]: `placeholderSrc` is actually static, we don't need `useRef` here

File: apps/cow-fi/pages/learn/[article].tsx [hunk 8]
```diff
@@ -119,6 +116,29 @@ function getRandomArticles(articles: Article[], count: number): Article[] {
   return shuffled.slice(0, count)
 }
 
+export function ArticleSharedRichTextComponent({ sharedRichText }: { sharedRichText: SharedRichTextComponent }) {
+  const { replaceImageUrls, LazyImage } = useLazyLoadImages()
+
+  // Preprocess the HTML content
+  const processedContent = React.useMemo(() => {
```
Comment (shoom3301) [inline apps/cow-fi/pages/learn/[article].tsx:123]: Just to be consistent with the rest of code-base, please use it as `import { useMemo } from 'react'`

**Review by elena-zh — COMMENTED** [review]
Hey @fairlighteth , everything is great!
However, there is one new thing appeared with the main menu in a mobile view:
 - white periods are displayed from the right side of every line
 - and the menu became scrollable (a bit) while it has fixed position on Prod
 
image
image

**Comment by fairlighteth** [comment]
@shoom3301 I should have addressed your comments.
@elena-zh the dots should be gone, could you double check (I could not re-produce myself). As for scrolling I couldn't reproduce either.

Did some refactoring for the CategoryLinks

**Comment by elena-zh** [comment]
The build has failed :(

**Comment by fairlighteth** [comment]
@elena-zh pushed a fix for that, please review

**Pull request merged**